File: Min_Stack.cpp
class MinStack {
private:
    stack<int> min_stack;
    stack<int> data;
public:
    MinStack() {
        // do initialization if necessary
    }

    void push(int number) {
        // write your code here
        if (min_stack.empty() || min_stack.top() >= number)
        {
            min_stack.emplace(number);
        }
        data.emplace(number);
    }

    int pop() {
        // write your code here
        int number = data.top();
        if (number == min_stack.top())
        {
            min_stack.pop();
        }
        data.pop();
        return number;
    }

    int min() {
        // write your code here
        if (!min_stack.empty())
        {
            return min_stack.top();
        }
    }
};

File: Min Stack.py
class MinStack(object):

    def __init__(self):
        # do some intialize if necessary
        self.mins = []
        self.data = []

    def push(self, number):
        # write yout code here
        self.data.append(number)
        if not self.mins or self.mins[-1] >= number:
            self.mins.append(number)

    def pop(self):
        # pop and return the top item in stack
        val = self.data.pop()
        if val == self.mins[-1]:
            self.mins.pop()
        return val

    def min(self):
        # return the minimum number in stack
        return self.mins[-1]